src/main/java/com/datastax/oss/doclet/ApiPlumber.java

```diff
@@ -15,179 +15,258 @@
-                  classTypeName,
-                  typeArgumentName,
-                  interfaceTypeName);
-            }
-          }
+      // skip interfaces and superclasses if marked with java-doc tag
+      if (!hasTag(docTrees.getDocCommentTree(clazz), EXCLUDE_TAG_NAME)) {
+        // check interfaces
+        for (TypeMirror interfaceTypeMirror : clazz.getInterfaces()) {
+          checkSuperType(typeUtils, interfaceTypeMirror, classTypeName, "parent interface");
         }
-      }
 
-      Type superclassType = clazz.superclassType();
-      if (superclassType != null) {
-        String superclassTypeName = superclassType.qualifiedTypeName();
-        if (clazz.tags(EXCLUDE_TAG_NAME).length == 0) {
-          checkAllowed(
-              superclassTypeName,
-              "Type %s leaks %s (as a superclass)%n",
-              classTypeName,
-              superclassTypeName);
-
-          if (superclassType instanceof ParameterizedType) {
-            for (Type type : superclassType.asParameterizedType().typeArguments()) {
-              String typeArgumentName = type.qualifiedTypeName();
-              checkAllowed(
-                  typeArgumentName,
-                  "Type %s leaks %s (as a type argument of its superclass %s)%n",
-                  classTypeName,
-                  typeArgumentName,
-                  superclassTypeName);
-            }
-          }
+        // check superclass
+        TypeMirror superclassTypeMirror = clazz.getSuperclass();
+        if (superclassTypeMirror.getKind() != TypeKind.NONE) {
+          checkSuperType(typeUtils, superclassTypeMirror, classTypeName, "superclass");
         }
       }
 
-      for (FieldDoc field : clazz.fields()) {
-        if (field.tags(EXCLUDE_TAG_NAME).length == 0) {
-          String fieldTypeName = field.type().qualifiedTypeName();
-          checkAllowed(fieldTypeName, "Field %s leaks %s%n", field.qualifiedName(), fieldTypeName);
+      // check contained fields, methods & constructors
+      for (Element enclosed : clazz.getEnclosedElements()) {
+        if (!visibleOutsidePackage(enclosed)
+            || hasTag(docTrees.getDocCommentTree(enclosed), EXCLUDE_TAG_NAME)) {
+          continue;
         }
-      }
 
-      for (MethodDoc method : clazz.methods()) {
-        if (method.tags(EXCLUDE_TAG_NAME).length == 0) {
-          String methodName = method.qualifiedName();
-          String returnTypeName = method.returnType().qualifiedTypeName();
-          checkAllowed(
-              returnTypeName,
-              "Method %s leaks %s (as its return type)%n",
-              methodName,
-              returnTypeName);
-
-          checkParametersAllowed(method);
+        if (enclosed.getKind().isField()) {
+          VariableElement field = (VariableElement) enclosed;
+          TypeMirror fieldTypeMirror = field.asType();
+
+          if (isCheckableType(fieldTypeMirror)) {
```

Comment:
I wonder if we should consider parameterized types here, e.g. `List<ForbiddenType>`. This would have to be done for parameter and return types too.
(not specific to your changes, if it's an issue it's also present in the original code)

Reply:
Thanks! Yes that does appear to leak in both 1.x and 2.x version. I've been collecting missing leak checks here: https://datastax-oss.atlassian.net/browse/JAVA-3093 - I've added this one to the list